iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/control_flow_ops.mlir

```diff
@@ -58,6 +58,60 @@ vm.module @my_module {
 
 // -----
 
+vm.module @my_module {
+  // CHECK: func @my_module_call_[[IMPORTFN:[^\(]+]]
+  vm.import @imported_fn(%arg0 : i32) -> i32
+
+  // CHECK: func @my_module_call_imported_fn
+  vm.func @call_imported_fn(%arg0 : i32) -> i32 {
+    // CHECK-NEXT: %0 = emitc.call "EMITC_STRUCT_PTR_MEMBER"(%arg2) {args = [0 : index, #emitc.opaque<"imports">]} : (!emitc.ptr<!emitc.opaque<"my_module_state_t">>) -> !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>
+    // CHECK-NEXT: %1 = emitc.call "EMITC_ARRAY_ELEMENT_ADDRESS"(%0) {args = [0 : index, 0 : ui32]} : (!emitc.ptr<!emitc.opaque<"iree_vm_function_t">>) -> !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>
+    // CHECK-NEXT: %2 = "emitc.variable"() {value = #emitc.opaque<"">} : () -> i32
+    // CHECK-NEXT: %3 = emitc.apply "&"(%2) : (i32) -> !emitc.ptr<!emitc.opaque<"int32_t">>
+    // CHECK-NEXT: %4 = call @my_module_call_[[IMPORTFN]](%arg0, %1, %arg3, %3) : (!emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, !emitc.ptr<!emitc.opaque<"iree_vm_function_t">>, i32, !emitc.ptr<!emitc.opaque<"int32_t">>) -> !emitc.opaque<"iree_status_t">
+    %0 = vm.call @imported_fn(%arg0) : (i32) -> i32
+    vm.return %0 : i32
+  }
+}
+
+// -----
+
+vm.module @my_module {
+  vm.func @internal_fn(%arg0 : i32) -> i32 {
+    vm.return %arg0 : i32
+  }
+
+  // CHECK-LABEL: @my_module_call_internal_fn
+  vm.func @call_internal_fn(%arg0 : i32) -> i32 {
+    // CHECK-NEXT: %0 = "emitc.variable"() {value = #emitc.opaque<"">} : () -> i32
+    // CHECK-NEXT: %1 = emitc.apply "&"(%0) : (i32) -> !emitc.ptr<!emitc.opaque<"int32_t">>
+    // CHECK-NEXT: %2 = call @my_module_internal_fn(%arg0, %arg1, %arg2, %arg3, %1) : (!emitc.ptr<!emitc.opaque<"iree_vm_stack_t">>, !emitc.ptr<!emitc.opaque<"my_module_t">>, !emitc.ptr<!emitc.opaque<"my_module_state_t">>, i32, !emitc.ptr<!emitc.opaque<"int32_t">>) -> !emitc.opaque<"iree_status_t">
+    %0 = vm.call @internal_fn(%arg0) : (i32) -> i32
```

Comment:
There's a lot going on with this check (and the others in this file). Could it be split onto multiple lines, or maybe only check certain properties of the conversion instead of an exact match? Maybe some inline comments (like those in https://github.com/google/iree/blob/main/iree/compiler/Dialect/Stream/Transforms/test/pack_constants.mlir, for example) would help too.

https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices

Reply:
The readability should be improved now with some comments and line breaks. I didn't find good ways to reduce the exact matches for the tests, but used some string substitution.